publish rework + reconnect retry buffer + inline subscriptions

israellot opened the pull request:
Followed your idea @80dB 
Performance was not hit at all apparently.

israellot committed:
implement reconnect retry buffer

israellot committed:
improve writer timing for benchmark

israellot committed:
small latency improvment

1. if no other data segment in queue, write directly to socket
2. allow synchronous continuation on sender queue

israellot committed:
replace list walk with simple case switch when parsing messages.

this avoids the indirect delegate call

israellot committed:
minor msg parser reduction in asm instructions

israellot committed:
Serialize publish messages directly to buffer

Here I moved the serialization to the left side of the queue, avoiding having to rent memory and copy that memory to a buffer at the other end.
The only case where memory is rented now is if the message is larger than the predefined buffers. The overhead of renting memory to serialize smaller messages was a bottleneck.
Memory allocation on the send path is well controlled now, since even renting from the pool, at high volume scenarios, would end up allocating since the pool might be exhausted.
Another factor is that now different publishes only contend to acquire a write slot, but can serialize in parallel to the buffer, and even after the buffer enters the queue. The socket sender task will commit and seal the buffer at the last possible moment.

This increased throughput by ~2x at least.

israellot commented:
Ok, I ended up trying another approach to the publish path, not renting memory anymore but writing directly to buffers that rotate. I think it was a breakthrough, the library can now flush twice more messages with no allocations. 
It turns out that on high volume scenarios, the ArrayPool would be exhausted and it would end up allocating anyway. And for small buffers, like a 32 byte message, renting and returning is way more expensive than a simple new[]. 
Take a look and let me know your thoughts. 
I have some ideas for the read path as well, stay tuned.

![Screenshot 2022-08-29 085038](https://user-images.githubusercontent.com/840042/187200324-a1beedf1-7433-4c75-bb02-f13aeb66973d.png)

Da-Teach commented:
I'm currently swamped with work [have an important demo in thursday] but I'll look into this asap.

israellot committed:
remove delegate allocations for publish method

israellot committed:
add inline message subscription processing

israellot commented:
Added one more proposal here. Inline subscriptions. 
Those are subscriptions that are processed directly in the sequence reader loop, without any extra copying of data. 
This adds huge performance gains for use cases where messages are small. You'll notice on benchmark that as messages get longer, the performance gap diminishes. It can now ingest ~3.5M messages/sec using this technic. Of course, that's something of a ceiling, as the benchmark just reads a long directly from the payload data, but it gives us an idea of how much overhead can be avoided. 
I added the subscribe with *Unsafe sufix, although I believe it can be made completely safe since the delegate argument is a readonly ref struct ( can't hold reference to it ), and the only concern would be holding a reference to the payload's ReadOnlySequence<byte>, which could be made a Span<byte> for most cases where the payload received is not split among multiple spans. 
Using this, roundtrips benchmarks have almost no allocations going on, besides the async state machine for the occasional awaits. 

![image](https://user-images.githubusercontent.com/840042/187953777-cf064eb3-5feb-434a-a77f-45a6dea39d26.png)

80dB commented:
I'm currently looking through the changes. One thing I noticed iss that while "Inline" is (in some cases much) faster, the "Normal" seems slower.  So looking into that a bit.

Old performance:
````
---Roundtrip raw pub 1 sub---
Target flood            0 B     1 pub   1 sub   :       3471,9k msg/s   6,94 MB/s Latency: 0,00ms 0,00ms 0,00ms
Target flood            8 B     1 pub   1 sub   :       3710,7k msg/s   10,39 MB/s Latency: 0,00ms 0,00ms 0,00ms
Target flood            16 B    1 pub   1 sub   :       2773,2k msg/s   10,26 MB/s Latency: 0,00ms 0,00ms 0,00ms
Target flood            32 B    1 pub   1 sub   :       2883,1k msg/s   15,28 MB/s Latency: 0,00ms 0,00ms 0,00ms
Target flood            64 B    1 pub   1 sub   :       2873,6k msg/s   24,43 MB/s Latency: 0,00ms 0,00ms 0,00ms
Target flood            128 B   1 pub   1 sub   :       3664,8k msg/s   54,97 MB/s Latency: 0,00ms 0,00ms 0,00ms
Target flood            256 B   1 pub   1 sub   :       2883,0k msg/s   80,15 MB/s Latency: 0,00ms 0,00ms 0,00ms
Target flood            512 B   1 pub   1 sub   :       2266,2k msg/s   121,02 MB/s Latency: 0,00ms 0,00ms 0,00ms
Target flood            1024 B  1 pub   1 sub   :       1417,1k msg/s   148,51 MB/s Latency: 0,00ms 0,00ms 0,00ms
```` 

New performance:
````
---Roundtrip raw pub 1 sub---
Target flood     (Normal)       0 B     1 pub   1 sub : 2792,0k msg/s   5,58 MB/s Lat: 0,00ms 0,00ms 0,00ms
Target flood     (Normal)       8 B     1 pub   1 sub : 3044,4k msg/s   8,52 MB/s Lat: 0,00ms 0,00ms 0,00ms
Target flood     (Normal)       16 B    1 pub   1 sub : 2896,0k msg/s   10,72 MB/s Lat: 0,00ms 0,00ms 0,00ms
Target flood     (Normal)       32 B    1 pub   1 sub : 3172,5k msg/s   17,13 MB/s Lat: 0,00ms 0,00ms 0,00ms
Target flood     (Normal)       64 B    1 pub   1 sub : 2761,9k msg/s   23,75 MB/s Lat: 0,00ms 0,00ms 0,00ms
Target flood     (Normal)       128 B   1 pub   1 sub : 2898,3k msg/s   43,76 MB/s Lat: 0,00ms 0,00ms 0,00ms
Target flood     (Normal)       256 B   1 pub   1 sub : 2393,9k msg/s   66,79 MB/s Lat: 0,00ms 0,00ms 0,00ms
Target flood     (Normal)       512 B   1 pub   1 sub : 1956,3k msg/s   104,66 MB/s Lat: 0,00ms 0,00ms 0,00ms
Target flood     (Normal)       1024 B  1 pub   1 sub : 1193,0k msg/s   125,03 MB/s Lat: 0,00ms 0,00ms 0,00ms
````

Some discrepancies but it seems that the "old" normal is faster. 

No doubt about Publish though, it's definitely faster.   Did you look at serializing the message in the "WriteSocketAsync" task vs the current NatsPublishBuffer/Channel implementation?

israellot commented:
Yes, I did try serializing after dequeue. The issue is mostly for smaller messages. As we can’t guarantee the payload buffer the user supplied will be valid, we need to either copy it or make the user wait for serialization with a task completion. That combined with the overhead of queuing that many messages slows down the entire process. Actually we can probably squeeze even more if we keep buffers thread local and publish from multiple threads sacrificing ordering, but that’s pure benchmark optimization imho.

israellot commented:
Regarding the read performance issue, yes, for smaller messages any extra overhead ( like an extra branch in code path ) ends up affecting performance. One experiment I did for example, is to drop the rented memory approach for small messages and let GC take care of cleanup. It actually doubles msg/sec, at the expense of more GC, but no explosions. I believe looking at those numbers at face value is not the right way to evaluate though. If you have a system throwing 2M / msgs per second at you in a single topic you will probably want to avoid a 100ms GC pause. My aim was actually to remove overheads, specially allocations, because those ultimately translates into GC pauses.
I guess my point is: if your use case is processing small messages in high volume, inline is the recommended approach. If your messages are bigger or are in sane volumes, it doesn’t matter much which approach you choose, both can handle, inline will offer you less allocations if you can examine the message without copying it. If you need to copy the payload anyway, then it’s irrelevant again which method you choose.

80dB commented:
> As we can’t guarantee the payload buffer the user supplied will be valid, we need to either copy it or make the user wait for serialization with a task completion. 

The only time we can't guarantee the payload buffer is when the user supplies his/her own byte[] , in every other instance [object or string] we can (and I suspect object to be the most common payload). Because with string the implicit conversion handles this and with an object the serialization from object to byte[] is also done by us.

The serialization from something to NatsPayload still needs to be done in the publish though, but the copy would only need to happen if "they" supplied their own NatsPayload [that is not based off a string].

israellot commented:
That's true. But if you think in terms of (asm operations/message byte), enqueuing and dequeing every message to a channel will necessarily degrade performance for smaller messages just due to the fact these operations add up, more jumps on while loops, allocations to hold elements in the queue, locks, etc. If the message size is big enough, eventually the cost is amortized and the approach won't matter much. Unfortunately, I didn't save the post queue serialization experiment I did, but shouldn't be hard to try it if you want to explore this avenue. 
I guess another benefit is that this approach let's serialization and buffer write to be done in a thread while the current buffer copy to the socket is happening in another one. This can probably be done solely on the writer task as well, but requires some task scheduling.

80dB commented:
I've played with queuing messages and like you expected, it's performance is lower than this branch. It's faster than the current implementation but slower than your implementation.

I did have a version that implemented a byte[]-pool and BlockingCollection, it was faster with <128b messages [11.2m vs 6.7m on 0-byte PUB's and 7.8m vs 6.1m on 128-byte PUB's], but it's performance tanked at 1024b [and higher] messages. However it's CPU-usage was through the roof [probably due to the lack of proper async-handling]. It does show potential though :)

israellot commented:
Good! I'm curious about this BlockingCollection approach.
It's a tradeoff at these levels. If you are willing to sacrifice CPU, you can block threads, spin lock, busy wait, etc to write to buffers as quickly as possible. I believe maxing out these benchmarks only for the sake of higher numbers is not a good approach in general. Also, to reach that throughput, messages have to be fed in a very specific way, practically a while(1) loop with doing nothing else. To be fair, such use cases are almost never real. Benchmarks help us measure how much overhead the library is introducing to the system, but the final number is almost never the operational point of anyone. 
Locking and contention is a whole science on its own, overhead and cpu usage increases exponentially as contention grows. 
It is also a compromise which message size to optimize for. If you have a system that you know only sends 8b messages, there's definitely a specific implementation that has optimal performance for that, and if you have a system that only sends 1kb messages, there's another one. I doubt those two will ever collide though. As a general library, the implementation should strike a balance across the whole range of possible use cases, as a best effort, while standing less in the way as possible.

80dB commented:
This is the high-cpu version:
https://github.com/80dB/AsyncNats/tree/feature/fast-high-cpu
It can probably be adopted to use less CPU by, for example, using ManualResetEvent (or a TaskCompletionSource - but that's a bit harder).

Pretty sure raw-message sending is broken in that branch.

israellot commented:
Some comments on your branch

```
public void Serialize(INatsClientMessage message)
        {
            while (true)
            {
                lock (_syncLock)
                {
                    var remaining = _bufferLength - _writeBufferLength;
                    if (message.Length < remaining)
                    {
                        message.Serialize(_writeBuffer.AsSpan(_writeBufferLength));
                        _writeBufferLength += message.Length;
                        _writeMessagesCount++;
                        break;
                    }

                    _sendQueue.Add((_writeBuffer, _writeBufferLength, _writeMessagesCount));

                    _writeBuffer = _pool.Take();
                    _writeBufferLength = 0;
                    _writeMessagesCount = 0;
                }
            }
        }
```
Here **message.Serialize** is called inside the lock, for larger messages, the lock will be held for longer and contention will hit earlier. How did it perform with multiple producers? I expect some complications. 
If *_sendQueue.Add* blocks because the queue is full, we end up doing sync wait over async, inside a lock, and the thread pool will probably need to spawn more threads. I guess those are the problems that lead to the creation of `Channel<T>`.

```
private async Task WriteSocketAsync(Socket socket, CancellationToken disconnectToken)
        {
            _logger?.LogTrace("Starting WriteSocketAsync loop");

            /*var reader = _senderChannel.Reader;
            var buffer = new byte[2 * 1024 * 1024];
            var socketBufferLength = socket.SendBufferSize;*/

            await SendConnect(socket, disconnectToken);
            await Resubscribe(socket, disconnectToken);
            while (!disconnectToken.IsCancellationRequested)
            {
                var (buffer, length, messageCount) = _buffer.GetReadBuffer();
                if (length == 0)
                {
                    await Task.Delay(10);
                    continue;
                }

```

The issue I see here is that this Task.Delay(10) will add unwanted latency for low frequency producers. Timers in .net actually have very poor resolution as well, and I believe the minimum delay possible is actually around 16ms.

80dB commented:
I doubt serializing within the lock is an issue, memory copy is (in general) really fast on a modern CPU.  I've experimented with a pre-serialize [serialize to a stackalloc byte[] and only doing a copy inside the lock] but had virtually no effect on performance. I actually suspect that lock contention is the biggest issue with multiple publishers. Because increasing the number of publishers decreases the throughput [100 publishers is slow as f...] - which can't be explained by the work inside the lock.

And reason there's a sleep in the send-loop is because I initially used a Yield but it made the CPU go to 100% 😆 The branch was more an experiment to see what idea's I could think off. It's not very good trade-offs.

That said, I suspect a non-Task based single publisher NatsConnection could be made to be a lot faster than the current "one connection fits all" strategy. And a single subscriber NatsConnection could probably be a lot faster as well. And just use 'blocking' calls instead of async calls. And threads instead of tasks. But none of those idea's "fit" the async-direction that C# / .NET is taking.

Anyways, I'll start reviewing this PR. The changes in this PR will make AsyncNats much faster than the other .NET libraries out there.

israellot commented:
The official NATS client has a sync implementation with dedicated threads. I guess we can expect similar results on that path. 
It's not bad per se, but it comes with some overheads and tradeoffs of additional long running threads, locks , etc. 
I believe the beauty of the async implementation is propagating backpressure, slowing down producers instead of sync blocking threads. Today it's almost certain people are going to be publishing from a Thread Pool thread, and mixing it with sync code is detrimental to the overall system. 
To be honest, for my use case, I'm much more concerned about consistent low latency rather than throughput. I need a system with minimal GC pauses and artificial delays. And besides that, we can easily push higher throughput now by sending over multiple physical connections, it's a common strategy.

israellot commented:
@80dB Do you think it's possible for us to evaluate this PR and merge what makes sense and leave out what doesn't?

80dB commented:
Yes, I've been extremely busy and I didn't get around to this. I'll see what I can do this weekend and/or early next week.

israellot commented:
Just pinging this topic :)

80dB commented on AsyncNats/Messages/NatsMessageParser.cs:
```diff
@@ -2,80 +2,407 @@
             return messageCount;
         }
+
+        public NatsMsg? ParseMessage(in ReadOnlySpan<byte> line, ref SequenceReader<byte> reader, out bool inlined)
```
I'm not a huge fan of having these message parsers moved into the NatsMessageParser - Can we move them to their own classes (like NatsMsgParser or something) ?

israellot replied:
> I might look at it again at a later point [have a headache], but overall looks good. Minor points here and there.
> 
> It might be an idea to somehow clone the InlineSubscriptions from a concurrentdictionary to a normal dictionary (and use an overwrite strategy) after a change. Concurrent dictionaries are much heavier to read compared to normal dictionaries (that said I suspect even a "hasInlineSubs" check would be nice for a normal dictionary).

That's ok, let me take a run on your comments and suggestions.

israellot replied:
I'm not a fan too. The reason here was to avoid the indirection of calls made via interface methods, and potentially allow inlining, since the message parser method ( the most used and critical ) now access the reference dictionary of inline subscriptions.
Let me know if I should explore alternatives, I'm leaving as is for now.

80dB commented on AsyncNats/Util/NatsKey.cs:
```diff
@@ -3,10 +3,29 @@
     using System.Runtime.CompilerServices;
     using System.Text;
 
+    public readonly ref struct NatsInlineKey
```
Should be moved to its own file

80dB replied:
And how big is the advantage of a ref-struct vs the normal struct? [more: do we need the aditional NatsInlineKey struct?]

israellot replied:
At this level, the advantage is small, but not zero. It avoids a potential copy instruction when passing it around method calls.   It also has to be a ref readonly struct in order to contain a ReadOnlySpan<byte> instead of a ReadonlyMemory<byte> as the normal NatsKey does, allowing us to use stack allocated keys.

80dB commented on AsyncNats/Util/NatsPublishChannel.cs:
```diff
@@ -0,0 +1,122 @@
+        public async ValueTask PublishInternal<T>(T msg, CancellationToken cancellationToken) where T : INatsClientMessage
+        {
+
+        retry:
```
Labels are really "not done", can we change this into a while?

israellot replied:
Done

80dB commented on AsyncNats/Util/NatsPublishChannel.cs:
```diff
@@ -0,0 +1,122 @@
+        public ChannelWriter<NatsPublishBuffer> Writer => _channel.Writer;
+
+
+        readonly Channel<NatsPublishBuffer> _channel;
```
Missing visibility keyword [private]

80dB replied:
Only mentioned it here but it's in several classes, explicit declaration of visibility is preferred

israellot replied:
Right, ran code cleanup adding access modifier to all fields.

80dB commented on AsyncNats/Util/NatsPublishChannel.cs:
```diff
@@ -0,0 +1,122 @@
+    {
+        public int DefaultBufferLength { get; set; } = 1024 * 64;
+
+        readonly struct NatsPublishAction
```
This structure isn't used?

israellot replied:
Right. Leftover from some other approach.

80dB commented on AsyncNats/Messages/NatsMessageParser.cs:
```diff
@@ -2,80 +2,407 @@
+            long sid = 0;
+            pointer = (splitCount == 1) ? headerSizeEnd : splits[0] - 1;
+            multiplier = 1;
+            currentByte = Unsafe.Add(ref lineRef, pointer);
```
This is faster than Utf8Parser.TryParse?

israellot replied:
It is in this particular case because digit count is known and we assume bytes are in the right format, avoiding some bound checks.

80dB commented on AsyncNats/Messages/NatsMessageParser.cs:
```diff
@@ -2,80 +2,407 @@
+                currentByte = Unsafe.Add(ref lineRef, pointer);
+            } while (currentByte != ' ');
+
+            if (_inlineSubscriptions.TryGetValue(sid, out var inlineSubscription))
```
There's a big chance (when looking at reference code) that having a boolean "hasInlineSubscriptions" can have a positive effect as TryGetValue does a lot of work for (event) empty dictionaries. See: https://github.com/microsoft/referencesource/blob/master/mscorlib/system/collections/Concurrent/ConcurrentDictionary.cs#L498

israellot replied:
You are right. The issue here is that the dictionary is mutable.  There's probably room for improvement by using an immutable dictionary and swapping references. I feel it can be addressed in a second pass.

80dB commented on AsyncNats/NatsConnection.cs:
```diff
@@ -34,10 +32,14 @@ public class NatsConnection : INatsConnection
-        private readonly Channel<IMemoryOwner<byte>> _senderChannel;
+        private ReadOnlyMemory<byte> _reconnectResendBuffer = ReadOnlyMemory<byte>.Empty;
+
+        //private readonly Channel<NatsMemoryOwner> _senderChannel;
```
Remove commented code

80dB commented:
I might look at it again at a later point [have a headache], but overall looks good.  Minor points here and there.

It might be an idea to somehow clone the InlineSubscriptions from a concurrentdictionary to a normal dictionary (and use an overwrite strategy) after a change. Concurrent dictionaries are much heavier to read compared to normal dictionaries (that said I suspect even a "hasInlineSubs" check would be nice for a normal dictionary).

israellot committed:
PR cleanup

israellot commented:
PR updated

80dB committed:
Merge branch 'master' into pr/10

80dB committed:
Small change (.NET3.1 -> .NET6)
